File: C++/70. Climbing Stairs.cpp
#include <bits/stdc++.h>

using namespace std;

// Basic Solution
int climbStairs_native(int n)
{
	if (n == 1 or n == 2)
		return n;
	return climbStairs_native(n - 1) + climbStairs_native(n - 2);
}

// By using Top down approach
int climbStairs(int n, vector<int> &dp)
{
	if (n == 1 or n == 2)
		return n;

	if (dp[n] == 0)
		dp[n] = climbStairs(n - 1) + climbStairs(n - 2);

	return dp[n];
}

int climbStairs(int n)
{
	vector<int> dp(n + 1, 0);
	return climbStairs(n, dp);
}

// By using bottom up approach
int climbStairs(int n)
{
	if (n == 1 or n == 2)
		return n;

	vector<int> dp(n + 1, 0);
	dp[1] = 1;
	dp[2] = 2;

	for (int i = 3; i <= n; i++)
		dp[i] = dp[i - 1] + dp[i - 2];

	return dp[n];
}

// Optimizing the Space complexity
int climbStairs(int n)
{
	if (n == 1 or n == 2)
		return n;

	int prev1 = 1, prev2 = 2;
	int cur;

	for (int i = 3; i <= n; i++)
	{
		cur = prev1 + prev2;
		prev1 = prev2;
		prev2 = cur;
	}
	return cur;
}

int climbStairs(int n)
{
	int a = 0, b = 1, c = 1;

	for (int i = 1; i < n; i++)
	{
		a = b;
		b = c;
		c = a + b;
	}
	return c;
}

File: Java/70. Climbing Stairs.java
class Solution {
    /*
     * Brutte-Force Approach
     * Time-Limit-Exceed
     * Time: O(2^n)
     * Space: O(n)
     */
    public int climbStairs_1(int n) {
        if (n <= 3)
            return n;
        return climbStairs(n - 1) + climbStairs(n - 2);
    }

    /*
     * DP Solution Top Down Solution
     * Time: O(n)
     * Space: O(n)
     */
    public int climbStairs_2(int n) {
        if (n <= 2)
            return n;

        int[] cache = new int[n + 1];
        cache[1] = 1;
        cache[2] = 2;

        return climbStairsUtil(n, cache);
    }

    private int climbStairsUtil(int n, int[] cache) {
        if (cache[n] == 0)
            cache[n] = climbStairsUtil(n - 1, cache) + climbStairsUtil(n - 2, cache);

        return cache[n];
    }

    /*
     * DP Bottom UP Solution
     * Time: O(n)
     * Space: O(n)
     */
    public int climbStairs_3(int n) {
        if (n <= 2)
            return n;

        int[] cache = new int[n + 1];
        cache[1] = 1;
        cache[2] = 2;

        for (int i = 3; i <= n; i++)
            cache[i] = cache[i - 1] + cache[i - 2];

        return cache[n];
    }

    /*
     * Space Optimal Solution
     * Time: O(n)
     * Space: O(1)
     */
    public int climbStairs(int n) {
        if (n <= 2)
            return n;

        int a = 1, b = 2, c = 0;

        for (int i = 3; i <= n; i++) {
            c = a + b;
            a = b;
            b = c;
        }
        return c;
    }
}